### DISCUSSION TITLE
Taking inspiration from scallop lang

### DISCUSSION THREAD

**Discussion by RAbraham** [cozodb/cozo#34] [Ideas]
Hi,
I wanted to start out by saying how excited I am to see this work. I have been a long time Datalog fan. 
One of the most exciting Datalog inspired languages for me other than cozodb is Scallop Lang: https://github.com/scallop-lang/scallop
It's a language for neuro-symbolic programming with integrations with Pytorch. It has probabilistic programming also. 

I only share this as I see in your 0.5 release, that you have some some intentions of integrating PyTorch. Scallop may be very far away from your roadmap but I thought I'll share it in case it gives you any inspiration.

I have no affiliation with this project :). 

Resources
- https://scallop-lang.github.io/
- https://www.youtube.com/watch?v=QBx8vt8NC2k ( A talk by the professor behind this language)

**Comment by zh217** [comment]
Thanks for the suggestion! I haven't heard of scallop before, but reading their documentation it seems that they support back propagation through Datalog rules? That's really amazing and is something that I haven't thought about before. I will definitely investigate further.

**Reply by RAbraham** [reply]
Yeah, I'm still digesting the idea myself but I believe there is back propagation happening(as per the youtube video).
Typically you have two components, the model followed by the rules. But contrary to typical machine learning, the labels are on the rules.
E.g. with MNIST, we typically have the labels as the number that the image represents(e.g. 4). But, imagine a hypothetical scenario where we don't have that but our labels are the sum of two images.
e.g. img1 + img2 = 5. We then model a rule `sum_2(a + b) :- digit_1(a), digit_2(b)`. 
We then have a two component architecture where the combination of the model and the rule above are trained to possibly provide better accuracy. It is not very clear in the paper itself(to my knowledge) but the paper it refers to (DeepProblog: https://arxiv.org/abs/1805.10872) has a very clear chart (Figure 3) indicating that such an approach has 1) higher accuracy and lower loss 2) that it reaches the optimal values at lower epochs! Well, that is my interpretation atleast.

ref: see section "Learning MNIST Sum-2" , https://scallop-lang.github.io/tutorial.html#section-11

**Reply by RAbraham** [reply]
Even if we don't use Scallop for machine learning, it's still a probabilistic datalog which is exciting by itself to model real world problems and it excites me to think of even just that in a production centric datalog like cozodb.

**Reply by RAbraham** [reply]
note that the probabilistic reasoning is activated by modes so it may be difficult to include that in a high volume/transaction environment that I understand cozodb is suited for.

**Reply by zh217** [reply]
Yes Cozo is supposed to be a high volume/transaction database, but depending on how "invasive" the probabilistic reasoning part is, it may be possible to integrate it inside Cozo. That's what the "fixed rule" and "callback/streaming" business in Cozo is for, and the recent introduction of custom fixed rules implementable in Python (as opposed to Rust-only) is mainly for integrating all kinds of machine learning / deep learning models _within_ Datalog rules (it already works quite well for integrating simple GNN models, but for production work we really need database replication, which won't be ready for quite a while). When I find some spare time, I will definitely try to see what will come out when scallop is integrated inside Cozo in this way.

**Reply by RAbraham** [reply]
That's great. Looking forward to it!
